### PR TITLE
Cache dom views

### LINKED ISSUE #129: Enhancement to main screen browsing experience.
This has been brought up a few times and I wanted to suggest an idea on how to resolve the problem.

**Problem:** Clicking an item on the homepage takes you to the contract detail screen on the vendor's page and then clicking back to home refreshes the entire list of items. Once the items reload again, the ordering of the items no longer matches the prior order, which makes the user have to dig through the same items multiple times and having to wait for it to reload feels unnecessary. This doesn't make for a great browsing experience, in fact it's pretty bad.

**Solution:** Let's identify when the user is casually checking out items; toggling back and forth between home and contract detail pages. If they are casually browsing, do not rebuild the list, instead cache it. We can safely append new items to the bottom of the list (lazy load style), but let's not blow away what has already been loaded.

At some point (let's say 30 minutes of inactivity), delete the cache and fetch from the network. Fetching should also happen on launch of the client and clicking reload while on the main screen could also destroy the cache and fetch from the network.

The only negative I can think of is nodes may go offline while they're still displaying in your cached list. I'd imagine this would only happen on a small percentages of clicks though since the cache doesn't live very long.

Open to ideas on how to make this even better.

### LINKED ISSUE #1570: Cache Views in DOM
- when the Discover and UserPage views are navigated away from, instead of removing the views, keep them in memory for a limited time
- a method of saving UserPage views by ID will need to be added, so more than one can be held in memory
- if the Discover or UserPage view is loaded again, and the limited time has not expired, show the existing view instead of loading data and re-rendering it (for the UserPage, only if the ID is the same)
- this is an experimental issue, it should be prototyped first to see if this approach is feasible

### PR DESCRIPTION
**Pull request by rmisio**
closes #129 
partially fixes #1570 

Applies view caching, including restoring of the scroll position, to the Discover page & the Settings page. This is not applied to the Userpage, because that view needs some refactoring to support it first (will look to do that a little later and will make a separate PR).

### CODE REVIEW

File: js/router.js [hunk 1]
```diff
@@ -159,43 +184,107 @@ module.exports = Backbone.Router.extend({
...
-      (loadingConfig = __.result(view, 'loadingConfig')) &&
-      loadingConfig.promise &&
-      typeof loadingConfig.promise.then === 'function') {
-      this.launchPageConnectModal(loadingConfig);
+    // let's update the cache of our existing view (if it's cached && not expired) so
+    // cachedAt is updated and the user has up until the view's 'cacheExpires' amount of
+    // time to come back to it in it's current state.
+    for (var key in this.viewCache) {
+      let cached = this.viewCache[key],
+          curHash = location.hash.slice(1);
```
Comment (jjeffryes) [inline js/router.js]: What is curHash used for? I can't find it anywhere else in the code.

Comment (rmisio) [inline js/router.js]: No idea - I'm sure it had a use at some point, but I must have changed the code and forgot to clean it up. I've removed it.

**Comment by jjeffryes** [comment]
When I merged this with master, the search bar on the discover stopped working right. When I typed in a search term and clicked enter, it didn't hide the Personalize/Random tab. Sometimes it would work, but the X button in the search bar wouldn't register clicks.

This might be due to the merge, or do to something that changed since you started on this. Can you bring this up to date with the master branch and check es-lint for any issues?

**Comment by rmisio** [comment]
Did a merge and cleaned some stuff up. Search bar seems to be working correctly now.

**Pull request merged**